File: DSAlgorithms/patterns/sliding_window/7_longest_substring_replacement.ts
const length_of_longest_substring = (str: string, k: number) =>
{
    let window_start = 0;
    let max_length = 0;
    let max_repeat_letter_count = 0;
    const frequency_map = new Map<string, number>();
    for(let window_end = 0; window_end < str.length; window_end++)
    {
        const right_char = str[window_end];
        if(!frequency_map.has(right_char))
            frequency_map.set(right_char, 0);
        let current = frequency_map.get(right_char);
        if(current !== undefined)
        {
            frequency_map.set(right_char, current + 1);
        }
        current = frequency_map.get(right_char);
        if(current !== undefined)
            max_repeat_letter_count = Math.max(max_repeat_letter_count, current);
        if((window_end - window_start + 1 - max_repeat_letter_count) > k)
        {
            const left_char = str[window_start];
            const current = frequency_map.get(left_char);
            if(current !== undefined)
                frequency_map.set(left_char, current - 1);
            window_start++;
        }
        max_length = Math.max(max_length, window_end - window_start + 1);
    }
    return max_length;
}
console.log(length_of_longest_substring('aabccbb', 2));
console.log(length_of_longest_substring('abbcb', 1));
console.log(length_of_longest_substring('abccde', 1));

File: DSAlgorithms/patterns/sliding_window/7_longest_substring_replacement.cpp
#include <iostream>
#include <bits/stdc++.h>
#include <unordered_map>

class char_replacement 
{
    public: 
        static int find_length(const std::string &str, int k)
        {
            int l = 0;
            int longest = 0;
            int max_repeated_letter_count = 0;
            std::unordered_map<char, int> letter_freq;
            for(int r = 0; r < str.length(); r++)
            {
                char right_char = str[r];
                letter_freq[right_char]++;
                max_repeated_letter_count = std::max(max_repeated_letter_count, letter_freq[right_char]);
                if(r - l + 1 - max_repeated_letter_count > k)
                {
                    char left_char = str[l];
                    letter_freq[left_char]--;
                    l++;
                }
                longest = std::max(longest, r - l + 1);
            }
            return longest;
        }
};

int main(int argc, char *argv[])
{
    std::cout<<char_replacement::find_length("aabccbb", 2)<< '\n';
    std::cout<<char_replacement::find_length("abbcb", 1)<< '\n';
    std::cout<<char_replacement::find_length("abccde", 1)<< '\n';
}
